src/Bindings/LuaStateParams.h

```diff
@@ -0,0 +1,571 @@
+		Item->DoSomething(Param1, Param2, Param3);
+		break;
+	}
+	case 1:
+	{
+		// The second overload was used; Item, ParamA and ParamB are valid
+		Item->DoSomething(ParamA, ParamB);
+		break;
+	}
+	// No need for default: if no overloads match, the function raises a Lua error, returning to the Lua's caller
+}
+*/
+class cLuaStateParams
+{
+public:
+	/** A wrapper used in cLuaStateParams::Read() to signalize that the value must be a non-nil pointer when read from Lua. */
+	template <typename T>
+	class cNonNil
+	{
+		friend class cLuaStateParams;
+		typedef T * Tptr;
+	public:
+		typedef T ParentType;
+
+		cNonNil(Tptr & a_Dest):
+			m_Dest(a_Dest)
+		{
+		}
+
+	protected:
+		/** The destination pointer where the param will be written. */
+		Tptr & m_Dest;
+	};
+	template <typename T> static cNonNil<T> nonNil(T * & a_Dest) { return cNonNil<T>(a_Dest); }
+
+
+	/** A wrapper used in cLuaStateParams::Read() to signalize that the value is a "self" - it must be a non-nil pointer when read from Lua,
+	and has a special error message when mismatched. */
+	template <typename T>
+	class cSelf
+	{
+		friend class cLuaStateParams;
+		typedef T * Tptr;
+	public:
+		typedef T ParentType;
+
+		cSelf(Tptr & a_Self):
+			m_Self(a_Self)
+		{
+		}
+
+	protected:
+		Tptr & m_Self;
+	};
+	template <typename T> static cSelf<T> self(T * & a_Self) { return cSelf<T>(a_Self); }
+
+
+	/** A wrapper used in cLuaState::ReadParams() to signalize that the value is a "static self",
+	it must be the table representing the class T when read from Lua. */
+	template <class T>
+	class cStaticSelf
+	{
+	public:
+		typedef T ParentType;
+	};
+	template <class T> static cStaticSelf<T> staticSelf() { return cStaticSelf<T>(); }
+
+
+
+protected:
+
+	/** Attempts to match the params on the Lua stack to the API function overloads given (recursively).
+	a_CurOverload is the index of the currently processed overload (recursion level),
+	Returns the index of the overload that matches the parameters.
+	If no overloads match, returns -1. */
+	template <typename... T1, typename... T2>
+	static int ReadInternal(cLuaState & a_LuaState, int a_CurOverload, std::tuple<T1...> & a_Ovl, T2 &&... a_OtherParams)
+	{
+		// Try to read this overload
+		if (ReadSingleOverload(a_LuaState, a_Ovl))
+		{
+			return a_CurOverload;
+		}
+
+		// Try the next overload:
+		return ReadInternal(a_LuaState, a_CurOverload + 1, std::forward<T2>(a_OtherParams)...);
+	}
+
+
+	/** Terminator for the template-based recursion of the function above - for a single overload. */
+	template <typename... T1>
+	static int ReadInternal(cLuaState & a_LuaState, int a_CurOverload, std::tuple<T1...> & a_Ovl)
+	{
+		// Try to read this overload
+		if (ReadSingleOverload(a_LuaState, a_Ovl))
+		{
+			return a_CurOverload;
+		}
+
+		// No more overloads, report failure
+		return -1;
+	}
+
+
+
+	/** Helper struct to implement iterating over std::tuple elements */
+	template <int N> struct SizeT {};
```

Comment:
`std::integral_constant<size_t, N>`?

Reply:
Just changing the type to `size_t` seemed to help in my tests